discovery_engine_core/core/src/engine.rs

```diff
@@ -589,52 +594,48 @@ async fn update_stacks<'a>(
         .then(|| ranker.take_key_phrases(take_top))
         .unwrap_or_default();
 
+    // Here we gather new documents for all relevant stacks, and put them into a vector.
+    // We don't update the stacks immediately, because we want to de-duplicate the documents
+    // across stacks first.
     let mut ready_stacks = stacks.len();
     let mut errors = Vec::new();
+    let mut all_documents: Vec<Document> = Vec::new();
     for stack in &mut stacks {
-        let articles = match stack.new_items(&key_phrases, history).await {
-            Ok(articles) => articles,
-            Err(stack::Error::New(NewItemsError::NotReady)) => {
+        let maybe_new_documents =
+            fetch_new_documents_for_stack(stack, ranker, &key_phrases, history).await;
+
+        match maybe_new_documents {
+            Err(Error::StackOpFailed(stack::Error::New(NewItemsError::NotReady))) => {
                 ready_stacks -= 1;
                 continue;
             }
             Err(error) => {
-                let error = Error::StackOpFailed(error);
                 error!("{}", error);
                 errors.push(error);
                 continue;
             }
+            Ok(documents) => all_documents.extend(documents),
         };
 
-        let id = stack.id();
-        let articles_len = articles.len();
-        let (documents, articles_errors) = articles
-            .into_par_iter()
-            .map(|article| {
-                let embedding =
-                    ranker
-                        .compute_smbert(article.excerpt_or_title())
-                        .map_err(|error| {
-                            let error = Error::Ranker(error);
-                            error!("{}", error);
-                            error
-                        })?;
-                document_from_article(article, id, embedding).map_err(|error| {
-                    let error = Error::Document(error);
-                    error!("{}", error);
-                    error
-                })
-            })
-            .partition_map::<Vec<_>, Vec<_>, _, _, _>(|result| match result {
-                Ok(document) => Either::Left(document),
-                Err(error) => Either::Right(error),
-            });
-        // only push an error if the articles aren't empty for other reasons and all articles failed
-        if articles_len > 0 && articles_errors.len() == articles_len {
-            errors.push(Error::Errors(articles_errors));
-            continue;
-        }
+        // Since we need to de-duplicate not only the newly retrieved documents among themselves,
+        // but also consider the existing documents in the stacks, we extract them into
+        // `all_documents` here.
+        all_documents.extend(stack.drain_documents());
```

Comment:
Nice catch. We might need a second loop in general, because `stacks` here means only the "ready" stacks (first line of the function), but we would likely want to de-duplicate across all of them, not just the ready ones.

Reply:
Can't we run the dedup only at the end of update_stacks on all stacks? and before the dedup we update the stacks as before.

Reply:
We could do that, but in that case (at least currently) we'd be ranking the documents twice (on each update), which I suppose we want to avoid ❓